### PR TITLE
Implement the likes local tracking feature

### LINKED ISSUE #57: Track liked episodes locally
Partially blocked by #42. Use [`svelte-persistent-writable`](https://github.com/cdellacqua/svelte-persistent-writable).

**Acceptance criteria**:
- The likes are restored after a page reload within a single browser
- The like button works on the Now Playing page
- New likes are written to the `localStorage`
- Liked episodes are marked prominently on the **Podcast Episodes** page
- **Your Library** page displays the podcasts that have liked episodes

### PR DESCRIPTION
**Pull request by aabounegm**
Adds support for the like button to actually store liked episodes locally and display them in the library page.
Take care that this is my first piece of code to write in FSD, so keep an eye open to mistakes in that area.
The PR is marked as draft since the criterion "_The like button works on the Now Playing page_" is not yet met as that page is not yet constructed.

Closes #57

### CODE REVIEW

File: src/lib/features/like-episode/model/like.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import type { Episode } from '$lib/shared/api';
+import { persistentWritable, localStorageAdapter } from 'svelte-persistent-writable';
+
+/**
+ * The store containing set of episodes liked by the current user.
+ */
+export const likesStore = persistentWritable(new Map<Episode['id'], Episode>(), {
```
Comment (illright) [inline src/lib/features/like-episode/model/like.ts:7]: [question](https://conventionalcomments.org/#labels): are you sure we need to store the episodes here? I imagine we're going to have a separate store for episodes

File: src/lib/features/like-episode/model/like.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    },
+    deserialize(serialized: string) {
+      const episodes: Episode[] = JSON.parse(serialized);
+      return new Map(episodes.map((episode) => [episode.id, episode]));
+    },
+  }),
+});
+
+/**
+ * Register that the user liked the given episode.
```
Comment (illright) [inline src/lib/features/like-episode/model/like.ts:21]: [praise](https://conventionalcomments.org/#labels): thank you for documenting your code, one does not see that often in the wild ;)

File: src/lib/features/like-episode/model/like.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+export function addLike(episode: Episode) {
+  likesStore.update((likes) => likes.set(episode.id, episode));
+}
+
+/**
+ * Remove the like for the given episode.
+ * @param episode The id of the episode whose like is to be removed.
+ */
+export function removeLike(episode: Episode) {
+  likesStore.update((likes) => (likes.delete(episode.id), likes));
```
Comment (illright) [inline src/lib/features/like-episode/model/like.ts:33]: [issue](https://conventionalcomments.org/#labels) (non-blocking): however elegant, this snippet is exotic, not very familiar.

[suggestion](https://conventionalcomments.org/#labels) (non-blocking): could we gain some code clarity by being explicit and writing out a return here instead?

File: src/routes/library.svelte [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+<!-- Favorites -->
+<section class="flex flex-col p-4 gap-4 items-stretch">
+  <H1>Favorites</H1>
+  {#each episodes as episode (episode.id)}
+    <EpisodeCard {episode} playing={false} />
+  {/each}
+</section>
+
+<!-- Offline -->
```
Comment (illright) [inline src/routes/library.svelte:19]: issue: let's think of `routes` as being on the app level, ideally keeping as little page and composition logic here as possible. 

suggestion: I would suggest moving this code to `lib/pages/library/ui/library-page.svelte`

**Review by illright — CHANGES_REQUESTED** [review]
issue: it would also be great if you added some unit tests to go with your functionality

praise: you've done quite well in terms of FSD :)

File: src/lib/features/like-episode/model/like.ts (hunk 1 above)
Comment (aabounegm) [reply src/lib/features/like-episode/model/like.ts:7]: If you go through the commit history (in particular, https://github.com/aabounegm/cast/pull/66/commits/996f810507f43fa8e18cd5305316fcc42db84d76) you can see that I wanted to store only the episode ID at first. Then, when working on the "My library" page, I realized that we still don't have a store for the episodes with which I can use the IDs, so I changed it to store them as well to be able to display them directly in the favorites section of the library.

Comment (illright) [reply src/lib/features/like-episode/model/like.ts:7]: But do you intend to change it later to only store the IDs, when we get our episode store in order?

**Review by illright — CHANGES_REQUESTED** [review]

File: cypress/integration/like.spec.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+it('likes an episode then finds it in the library', () => {
+  cy.visitAndWaitForHydration('/podcasts/1');
+
+  // Find a card and give it a name
+  cy.get('[data-cy="episode-card"]').first().as('episodeCard');
```
[the remark this answers is not in the document]
Comment (illright) [reply cypress/integration/like.spec.ts:5]: praise: But certainly mad props for adding an integration test to begin with. I was thinking we'll have to go back to it at the end of the project

File: src/lib/features/like-episode/model/like.ts (hunk 1 above)
Comment (aabounegm) [reply src/lib/features/like-episode/model/like.ts:7]: Hopefully yes, or at least the service that fetches them from the server

File: src/lib/pages/library/ui/library-page.svelte [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+  $: episodes = Array.from($likesStore.values());
+</script>
+
+<!-- Sign in card -->
+
+<!-- Favorites -->
+<section class="flex flex-col p-4 gap-4 items-stretch" data-cy="favorites-section">
+  <H1>Favorites</H1>
+  {#each episodes as episode (episode.id)}
```
[the remark this answers is not in the document]
Comment (aabounegm) [reply src/lib/pages/library/ui/library-page.svelte:14]: Yes, I noticed that in the design, but I disagree that it "makes sense" (though I should've probably mentioned it somewhere). I believe it's very weird to mark episodes as favorites and then see the entire podcast and not the episode you marked. What if you mark 2 episodes as favorites? Do you still only see the podcast? And then do you have to search through the podcast to find those favorite episodes? And then is it not possible to unmark the episodes as favorite from that screen, or would unmarking the podcast unmark all of its episodes? I imagine people would use the favorites as a sort of quick access to a few episodes they really like, not to the podcasts in which they liked episodes.
Imagine if after starring a repo in GitHub you go to your "Starred repos" tab to find the users/organizations (owners of the repos) instead!

_TL;DR: That's rather unintuitive and I implemented it that way on purpose._

File: src/lib/features/like-episode/spec/like.spec.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { likesStore, addLike, removeLike, toggleLike } from '../model/like';
```
[the remark this answers is not in the document]
Comment (aabounegm) [reply src/lib/features/like-episode/spec/like.spec.ts:1]: Yeah, that was an oversight while I was trying to copy your structure.

File: cypress/integration/like.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (aabounegm) [reply cypress/integration/like.spec.ts:5]: I understand this and I agree, but I couldn't find a way to semantically identify the elements uniquely without also depending on the HTML structure. Like, if I `findByRole('p')`, which one should it pick, the title or duration? And there may not always be ARIA attributes that make sense. I also definitely didn't want to depend on the CSS classes. If you can suggest a better way to find the element semantically without using `data-cy`, I'd gladly implement it.

[the remark this answers is not in the document]
Comment (illright) [reply cypress/integration/like.spec.ts:5]: Here's a helpful [_timestamped_ video](https://youtu.be/F8LH9d9eN3M?t=2486) of a person testing a similar application in Cypress with the Testing Library.

TL;DW:
* There's no way to identify a container by role, so our best bet is to use the `findByTestId` (and we can configure the Testing Library to search for the `data-cy` attribute to avoid writing out the CSS selector like that).
* Once you get ahold of the container, you can use semantic queries, mimic the behaviour of a user — first, ensure that the text of the sought episode title is present in the card, then find a button (`role="button"`) that has a "like" in the name and click that

So basically, I propose the following changes: 
1. Configure Cypress Testing Library to use `data-cy` attributes for the `findByTestId` query (but if I'm being honest, I'd prefer `data-test-id`, it's much clearer to me)
2. Change the way we locate the button to a `findByRole` query
3. Avoid `expect(title).to.be.a('string');`, because that doesn't concern the user, instead use the query `findByText` and assert its existence

[the remark this answers is not in the document]
Comment (illright) [reply cypress/integration/like.spec.ts:5]: @VanishMax let me know what you think of this

File: cypress/integration/like.spec.ts [hunk 8]
```diff
@@ -2,23 +2,20 @@ it('likes an episode then finds it in the library', () => {
...
 
   // Find a card and give it a name
-  cy.get('[data-cy="episode-card"]').first().as('episodeCard');
+  cy.findAllByTestId('episode-card').first().as('episodeCard');
 
   // Grab the episode title
   cy.get('@episodeCard')
-    .get('[data-cy="episode-title"]')
-    .first()
+    .findByRole('heading', { level: 2 })
```
[the remark this answers is not in the document]
Comment (illright) [reply cypress/integration/like.spec.ts:9]: If you agree with these suggestions, I can implement these changes

File: src/lib/pages/library/ui/library-page.svelte [hunk 9]
```diff
@@ -9,7 +9,7 @@
 <!-- Sign in card -->
 
 <!-- Favorites -->
-<section class="flex flex-col p-4 gap-4 items-stretch" data-cy="favorites-section">
+<section class="flex flex-col p-4 gap-4 items-stretch" aria-label="favorites">
```
[the remark this answers is not in the document]
Comment (aabounegm) [reply src/lib/pages/library/ui/library-page.svelte:12]: Yeah, I just didn't want to give an `id` to an element in a component, but now I realized that it's not in an `#each` or even a reusable component (it's a page) so it's probably fine

File: src/lib/entities/episode/ui/episode-card-shell.svelte [hunk 10]
```diff
@@ -11,10 +11,10 @@
     'flex flex-col rounded-xl bg-background-slightly-higher text-white py-3 px-4',
     episode.favorite && 'border-b-4 border-pink-500 box-border'
   )}
-  data-cy="episode-card"
+  data-testid="episode-card"
 >
   <div class="flex items-center pb-1">
-    <p class="font-medium text-sm flex-1" data-cy="episode-title">{episode.title}</p>
+    <h2 class="font-medium text-sm flex-1">{episode.title}</h2>
```
[the remark this answers is not in the document]
Comment (aabounegm) [reply src/lib/entities/episode/ui/episode-card-shell.svelte:17]: It doesn't really have to be, but a title felt like it deserved to be a heading (it has a higher importance than the description for screen readers, for example), and it also helps differentiate between the description semantically without having to use `data-textid`

File: cypress/integration/like.spec.ts [hunk 11]
```diff
@@ -2,23 +2,20 @@ it('likes an episode then finds it in the library', () => {
   cy.visitAndWaitForHydration('/podcasts/1');
 
   // Find a card and give it a name
-  cy.get('[data-cy="episode-card"]').first().as('episodeCard');
+  cy.findAllByTestId('episode-card').first().as('episodeCard');
```
[the remark this answers is not in the document]
Comment (aabounegm) [reply cypress/integration/like.spec.ts:5]: That makes sense. But also instead of liking all of them, maybe we should like random episodes and check for them?

File: cypress/integration/like.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (aabounegm) [reply cypress/integration/like.spec.ts:9]: > it's kind of bleeding through implementation details

That's mostly true, but I couldn't for the life of me figure out how to extract the title in a generic way to use it in queries; the best I could do is the `.invoke('text')` below it.

> I imagine that the user will locate the card they want by the text content of the episode they're looking for instead of relying on the heading structure

I didn't really understand what you meant. Does "text content" mean the episode title or description? Because I can easily imagine users looking at the title first and then the description as a sort of "double check" that it's the correct episode. That's also why I thought the title should be a heading, having in mind screen readers going through them and the users suffering through long descriptions, not realizing they should skip every other paragraph to focus on the titles

File: src/lib/pages/library/ui/library-page.svelte (hunk 9 above)
[the remark this answers is not in the document]
Comment (aabounegm) [reply src/lib/pages/library/ui/library-page.svelte]: Actually, I remembered now after the test failed that the other reason was to identify it easily in the test. Still, feel free to fix the test

File: cypress/integration/like.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (illright) [reply cypress/integration/like.spec.ts:5]: I don't see much of a reason to differentiate, it happens automatically anyway. I understand that it increases the runtime of our tests, but oh well

File: src/lib/entities/episode/ui/episode-card-shell.svelte (hunk 10 above)
[the remark this answers is not in the document]
Comment (illright) [reply src/lib/entities/episode/ui/episode-card-shell.svelte:17]: I don't think there's a need to differentiate it semantically. Screen reader users will be looking for text, and they'll locate it either way. I don't think that a heading is appropriate in places where it doesn't precede textual content

File: cypress/integration/like.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (illright) [reply cypress/integration/like.spec.ts:9]: > I couldn't for the life of me figure out how to extract the title in a generic way

That video I've [linked in the comment above](https://github.com/aabounegm/cast/pull/66#discussion_r815292398) tackles this problem in an alternative way — starting with the data and then proceeding to elements. Basically, we would have a data fixture containing the sample podcasts that we will render and in the test we would iterate over that data and assert it in multiple places.

> Does "text content" mean the episode title or description?

We don't have episode descriptions though. That's why the title will be the one and the only thing that users will read

File: src/lib/entities/episode/ui/episode-card-shell.svelte (hunk 10 above)
[the remark this answers is not in the document]
Comment (aabounegm) [reply src/lib/entities/episode/ui/episode-card-shell.svelte:17]: I thought we would have episode description, but even if not, there is another `<p>` element below that displays the duration and download button. Screen readers will not differentiate between both `<p>`s.

File: cypress/integration/like.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (aabounegm) [reply cypress/integration/like.spec.ts:5]: I just thought it would be more human-like, because almost no one would like every single episode. If we also check that there are no episodes there that we didn't like (as we probably should), this will protect against possible bugs such as simply displaying all episodes there are.

File: cypress/integration/like.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (aabounegm) [reply cypress/integration/like.spec.ts:9]: > we would have a data fixture containing the sample podcasts

I thought about that, but I felt that this way we wouldn't care about the data source, and for example it could test Supabase as well when it is implemented. On another thought, you may be right that a fixture is the best solution here (otherwise, we cannot guarantee the existence of data)

File: src/lib/entities/episode/ui/episode-card-shell.svelte (hunk 10 above)
[the remark this answers is not in the document]
Comment (illright) [reply src/lib/entities/episode/ui/episode-card-shell.svelte:17]: And it doesn't matter, as they will read out the name and the user will immediately know they're in the right place

**Review by illright — APPROVED** [review]
